### PR TITLE
[form-factors] Add subtracted [KKvDZ:2022A] form factors

### PR DESCRIPTION
**Pull request by SKuerten**
First try on adding the subtractions in kkvdz form factors.
ToDo:
 - [x] Assign zeroes to subtraction point and subtraction polynomial.
 - [x] Add option for switching between subtracted and unsubtracted dispersion relation.

### CODE REVIEW

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 1]
```diff
@@ -48,8 +51,36 @@ namespace eos
...
+                 { UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this) },
+                 { UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this) },
+                 { UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this),
+                   UsedParameter(p[0.0],  *this) }}},
+        _c_rho{{{ UsedParameter(p[0.0],  *this),
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:67]: Hm. Can you distinguish the subtraction constants in the ``rho`` dispersion relation from those of the ``omega`` dispersion relation?

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 2]
```diff
@@ -126,57 +157,67 @@ namespace eos
...
     {
         double a_0(a[0]),a_1(a[1]),a_2(a[2]);
+        double c_0(c[0]),c_1(c[1]),c_2(c[2]);
         const double diff_z = _z_omega(k2) - _z_omega(0.0);
         const double m_omega = BToOmega::m_V, m_omega_sq = m_omega * m_omega;
 
-        return 1.0 / (1.0 - k2 / m_r_sq) * ( m_omega_sq / ( m_omega_sq - q2 - 1.0i * m_omega * _width_omega(q2) ) ) *
-                (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) );
+        return 1.0 / (1.0 - k2 / m_r_sq) * m_omega_sq / ( m_omega_sq - q2 - 1.0i * m_omega * _width_omega(q2) ) * ( (1.0 - switch_subtracted) *
+                (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) ) + switch_subtracted * (c_0 + c_1 * diff_z + c_2 * power_of<2>(diff_z) + s_0 *
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:169]: I think this is incorrect. IMHO, the "constant" part does not enter multiplied by the omega's Breit-Wigner.

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 3]
```diff
@@ -126,57 +157,67 @@ namespace eos
...
+                                                                  const std::array<UsedParameter, 3> & c, const UsedParameter & s_0) const
     {
         double a_0(a[0]),a_1(a[1]),a_2(a[2]);
+        double c_0(c[0]),c_1(c[1]),c_2(c[2]);
         const double diff_z = _z_rho(k2) - _z_rho(0.0);
         const double m_rho = BToRho::m_V, m_rho_sq = m_rho * m_rho;
 
-        return 1.0 / (1.0 - k2 / m_r_sq) * ( m_rho_sq / ( m_rho_sq - q2 - 1.0i * sqrt(q2) * _width_rho(q2) ) ) *
-                (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) );
+        return 1.0 / (1.0 - k2 / m_r_sq) * m_rho_sq / ( m_rho_sq - q2 - 1.0i * sqrt(q2) * _width_rho(q2) ) * ( (1.0 - switch_subtracted) *
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:182]: Ditto.

**Comment by mreboud** [comment]
I also noticed that you wrote KKVDZ with an uppercase V, it should be lowercase :)

**Comment by mreboud** [comment]
Ideally it would be nice to have the options here: https://github.com/eos/eos/blob/bdab45c6861c7222e439372de4642c1eab09c654/eos/form-factors/parametric-kkvdz2022.cc#L226

But this is something I have to do throughout EOS anyway...

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 4]
```diff
@@ -24,7 +24,10 @@ using namespace std::literals::complex_literals;
 
 namespace eos
 {
-    KKvDZ2022FormFactors::KKvDZ2022FormFactors(const Parameters & p, const Options &) :
+    KKvDZ2022FormFactors::KKvDZ2022FormFactors(const Parameters & p, const Options & o) :
+        opt_subtracted(o, "subtracted", { "off", "on"}, "off"),
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:28]: Inconsistent whitespace after ``"on"``.

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 5]
```diff
@@ -24,7 +24,10 @@ using namespace std::literals::complex_literals;
...
+        switch_subtracted(0.0),
+        _s_0(p[0.0],  *this),
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:30]: Inconsistent whitespace before ``*this``.

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 6]
```diff
@@ -48,8 +51,24 @@ namespace eos
                  UsedParameter(p[_par_name("rho_3_2")],  *this) },
                { UsedParameter(p[_par_name("rho_4_0")],  *this),
                  UsedParameter(p[_par_name("rho_4_1")],  *this),
-                 UsedParameter(p[_par_name("rho_4_2")],  *this) }}}
-    {
+                 UsedParameter(p[_par_name("rho_4_2")],  *this) }}},
+        _c_subtraction{{{ UsedParameter(p[0.0],  *this),
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:55]: I had no idea this worked. Why are you looking up a parameter by index. Even more so, why are you using a ``float`` as the index?

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 7]
```diff
@@ -126,57 +145,76 @@ namespace eos
...
         double a_0(a[0]),a_1(a[1]),a_2(a[2]);
         const double diff_z = _z_omega(k2) - _z_omega(0.0);
         const double m_omega = BToOmega::m_V, m_omega_sq = m_omega * m_omega;
+        const complex<double> omega_bw = m_omega_sq / ( m_omega_sq - q2 - 1.0i * m_omega * _width_omega(q2) );
 
-        return 1.0 / (1.0 - k2 / m_r_sq) * ( m_omega_sq / ( m_omega_sq - q2 - 1.0i * m_omega * _width_omega(q2) ) ) *
-                (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) );
+        return 1.0 / (1.0 - k2 / m_r_sq) * (
+               omega_bw * (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) ) * (1.0 - switch_subtracted) +
+               switch_subtracted * (q2 - s_0) / (s_0 - m_omega_sq) * omega_bw * (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) ) );
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:158]: Please check if the sign of the new term is correct. I suspect that the factor should read ``(q2 - s_0) / (m_omega_sq - s_0)``.

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 8]
```diff
@@ -126,57 +145,76 @@ namespace eos
...
+                                                                  const UsedParameter & s_0) const
     {
         double a_0(a[0]),a_1(a[1]),a_2(a[2]);
         const double diff_z = _z_rho(k2) - _z_rho(0.0);
         const double m_rho = BToRho::m_V, m_rho_sq = m_rho * m_rho;
+        const complex<double> rho_bw = m_rho_sq / ( m_rho_sq - q2 - 1.0i * sqrt(q2) * _width_rho(q2) );
+
+        return 1.0 / (1.0 - k2 / m_r_sq) * (
+               rho_bw * (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) ) * (1.0 - switch_subtracted) +
+               switch_subtracted * (q2 - s_0) / (s_0 - m_rho_sq) * rho_bw * (a_0 + a_1 * diff_z + a_2 * power_of<2>(diff_z) ) );
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:172]: See above.

File: eos/form-factors/parametric-kkvdz2022.cc [hunk 9]
```diff
@@ -126,57 +145,76 @@ namespace eos
...
+    }
+
+    double
+    KKvDZ2022FormFactors::_subtraction_polynomial(const double & k2, const std::array<UsedParameter, 3> & c) const
```
Comment (dvandyk) [inline eos/form-factors/parametric-kkvdz2022.cc:176]: That's a good idea, adding a new function for the subtraction polynomial.

**Review by dvandyk — CHANGES_REQUESTED** [review]
I doubt that you tested this PR.

 - There is no changeset introducing the new parameters for the new subtraction parameters.
 - You do not associate new parameters with your newly-introduced ``UserParameter`` objects.
 - I think there is a sign error in the subtracted dispersion relation for both rho and omega.

**Review by dvandyk — APPROVED** [review]

**Pull request merged**